sound/soc/sof/intel/hda-ipc.c

```diff
@@ -254,20 +252,62 @@ irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
 	if (irq_status == 0xffffffff)
 		goto out;
 
-	/* IPC message ? */
 	if (irq_status & HDA_DSP_ADSPIS_IPC) {
 		/* disable IPC interrupt */
 		snd_sof_dsp_update_bits_unlocked(sdev, HDA_DSP_BAR,
 						 HDA_DSP_REG_ADSPIC,
 						 HDA_DSP_ADSPIC_IPC, 0);
-		ret = IRQ_WAKE_THREAD;
+		ret = true;
 	}
 
 out:
-	spin_unlock(&sdev->hw_lock);
+	spin_unlock_irq(&sdev->hw_lock);
 	return ret;
 }
 
+/* is this IRQ for ADSP or SDW? - we only care about IPC and SDW here */
+irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+	struct sof_intel_hda_dev *hdev = sdev->pdata->hw_pdata;
+	int ret = IRQ_NONE;
+
+	sdev->irq_event = 0;
+
+	/* IPC message ? */
+	if (is_ipc_irq(sdev)) {
+		/* disable IPC interrupt at is_ipc_irq() */
+		ret = IRQ_WAKE_THREAD;
+		sdev->irq_event |= SOF_IRQ_IPC;
+	}
+	/* SDW message */
+	if (hdev->sdw && is_sdw_intel_irq(hdev->sdw)) {
+		/* disable SDW interrupt */
+		hda_sdw_int_enable(sdev, false);
+		ret = IRQ_WAKE_THREAD;
+		sdev->irq_event |= SOF_IRQ_SDW;
+	}
+
+	return ret;
+}
+
+irqreturn_t hda_dsp_irq_thread(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+	struct sof_intel_hda_dev *hdev = sdev->pdata->hw_pdata;
+
+	if ((sdev->irq_event & SOF_IRQ_IPC) || is_ipc_irq(sdev))
+		sof_ops(sdev)->irq_thread(irq, context);
+	if (sdev->irq_event & SOF_IRQ_SDW)
+		sdw_intel_thread(irq, hdev->sdw);
+	else if (hdev->sdw && is_sdw_intel_irq(hdev->sdw)) {
+		hda_sdw_int_enable(sdev, false);
+		sdw_intel_thread(irq, hdev->sdw);
```

Comment:
sorry. I don't understand the last two cases here. Aren't they exactly the same thing?

Also what's the point of using the irq_events since we will also test for is_ipc_irq and is_sdw_intel_irq() in the thread?

Reply:
> sorry. I don't understand the last two cases here. Aren't they exactly the same thing?

``if (sdev->irq_event & SOF_IRQ_SDW)`` means ``is_sdw_intel_irq()`` return true in the handler, and we already disable interrupt there. So we don't need to disable interrupt there. ``else`` means ``is_sdw_intel_irq()`` return false. So we test ``is_sdw_intel_irq()`` here again to make sure we will handle sdw interrupt if there is any. And we need to disable sdw interrupt here since we didn't disable it in the handler.

> Also what's the point of using the irq_events since we will also test for is_ipc_irq and is_sdw_intel_irq() in the thread?

We don't have to test is irq again in the thread If irq_event is set. i.e. We will only test is irq when the test result is false on irq handler.